### PR TITLE
Encryption in aes-gcm of large buffers not only by 16 bytes

### PR DESCRIPTION
**Pull request by Bizyroth**
## Description

Actually, in case we encrypt larges buffers (typically 1024 bytes buffer), we are forced to encrypt them 16 bytes by 16 bytes until the end of the buffer (except for the last block). I add some change to the gcm.h and gcm.c files to allow the encryption of larges buffers in a different way. We can cut our large buffer in more buffers with a different size than 16 bytes and encrypt them one by one. The result will be the same as if we encrypt our large buffer in one time.  

I had a problem when i changed the crypto library to mbedtls in the libsrtp of Cisco. One of the unit test encrypts a large buffer in one time then piece by piece (with random size) and compares the results. 

### Description of the algorithm

I add two variables in the mbedtls_gcm_context:
1. ectr_remain: the previous ectr we used to encrypt data
2. the size of the last buffer we encrypt modulo 16

In gcm.c:
1. Before encrypt the buffer, we check if last buffer was 16 bytes. If not we use the old ectr before update the context
2. If, after the buffer encryption, there is still bytes we dont encrypt, we doesnt update the context
(For example, we encrypt 3 bytes, then recall gcm_update for 5 bytes. There is still 8 bytes to encrypt before the update of the context)
3. Then we encrypt the buffer normally.
4. We save the ectr in ectr_remain in case we didn't see 16 bytes and we compute the remaining bytes to encrypt in data_remain

## Todos
- [ ] Documentation
- [ ] Changelog updated

## Additional comments
It's my first pull request. If it's not good enough, tell me the problem and i will do it again

### CODE REVIEW

**Comment by RonEld** [comment]
Hi @Bizyroth Thank you for your contribution!
I would like to confirm that you have signed the Contributor's Licensse Agreement, and that [this](https://os.mbed.com/users/Bizyroth/) is your Mbed account

**Comment by RonEld** [comment]
In addition, please resolve the CI failures.
You should revert the changes you did to the Makefiles. These are auto-generated by Clang, and should not be pushed. I believe that reverting the Makefile changes should resolve some of your CI failures. Some failures are GCM related, so I believe it is related to your change.
Before you push your changes, please run `tests/scripts/check-names.sh` to see what caused one of the CI failures

File: tests/suites/test_suite_cipher.function [hunk 1]
```diff
@@ -712,3 +712,75 @@ void check_padding( int pad_mode, char *input_str, int ret, int dlen_check )
         TEST_ASSERT( dlen == (size_t) dlen_check );
 }
 /* END_CASE */
+
+/* BEGIN_CASE depends_on:MBEDTLS_CIPHER_MODE_GCM */
+void check_gcm( int cipher_id, char *hex_key, char *hex_iv )
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.function]: I believe this is a test for special use case of short buffers. Please rename the function name

Comment (Bizyroth) [reply tests/suites/test_suite_cipher.function]: This function encrypts a 1024 bytes buffer in one time and then encrypts the same 1024 bytes buffer piece by piece.  So the test is not specially to short buffers. This test is here to show that we can encrypt large buffer piece by piece but not only 16 bytes by 16 bytes.

File: library/gcm.c [hunk 2]
```diff
@@ -370,8 +370,52 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
     }
 
     ctx->len += length;
-
     p = input;
+    if( length != 0 )
+    {
+        if( (size_t) ctx->data_remain >= length )
+        {
+            for( i = ( 16 - ctx->data_remain );
```
Comment (RonEld) [inline library/gcm.c]: can `data_remain` be > 16?

File: include/mbedtls/gcm.h [hunk 3]
```diff
@@ -50,6 +50,8 @@ typedef struct {
     unsigned char y[16];        /*!< Y working value */
     unsigned char buf[16];      /*!< buf working value */
     int mode;                   /*!< Encrypt or Decrypt */
+    unsigned char ectr_remain[16];/*!<last ectr for remaining data */
+    uint64_t data_remain;		/*!<size of remaining data */
```
Comment (RonEld) [inline include/mbedtls/gcm.h]: does `data_remain` need to be `uint64_t`? Why isn't it `size_t`?

File: library/gcm.c [hunk 4]
```diff
@@ -370,8 +370,52 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
     }
 
     ctx->len += length;
-
     p = input;
+    if( length != 0 )
+    {
+        if( (size_t) ctx->data_remain >= length )
```
Comment (RonEld) [inline library/gcm.c]: Please add comment for use case of this if statement

File: library/gcm.c [hunk 5]
```diff
@@ -370,8 +370,52 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
...
+                if( ctx->mode == MBEDTLS_GCM_ENCRYPT )
+                	ctx->buf[i] ^= out_p[pre_len];
+
+                pre_len++;
+            }
+            ctx->data_remain -= length;
+
+            return ( 0 );
+        }
+        else
```
Comment (RonEld) [inline library/gcm.c]: Please add comment for use case of this else statement

File: library/gcm.c [hunk 6]
```diff
@@ -386,22 +430,32 @@ int mbedtls_gcm_update( mbedtls_gcm_context *ctx,
...
+            out_p[i + pre_len] = ectr[i] ^ p[i];
             if( ctx->mode == MBEDTLS_GCM_ENCRYPT )
-                ctx->buf[i] ^= out_p[i];
+                ctx->buf[i] ^= out_p[i+pre_len];
         }
-
-        gcm_mult( ctx, ctx->buf, ctx->buf );
+        ctx->data_remain = 16 - use_len;
+        
+		if( ctx->data_remain == 0 )
```
Comment (RonEld) [inline library/gcm.c]: fix indentation. Did you use tabs here?

File: library/gcm.c [hunk 7]
```diff
@@ -437,6 +494,9 @@ int mbedtls_gcm_finish( mbedtls_gcm_context *ctx,
             tag[i] ^= ctx->buf[i];
     }
 
+    mbedtls_zeroize( ctx->ectr_remain, 16 );
```
Comment (RonEld) [inline library/gcm.c]: change 16 to `sizeof(ctx->ectr_remain)`

File: tests/suites/test_suite_cipher.function [hunk 8]
```diff
@@ -712,3 +712,75 @@ void check_padding( int pad_mode, char *input_str, int ret, int dlen_check )
...
+    TEST_ASSERT( 0 == mbedtls_cipher_setup( &ctx, cipher_info ) );
+    TEST_ASSERT( 0 == mbedtls_cipher_setkey( &ctx, key, key_len * 8, MBEDTLS_ENCRYPT ) );
+    
+    /* Prepare context for full encryption */
+    TEST_ASSERT( 0 == mbedtls_cipher_set_iv( &ctx, iv, iv_len ) );
+    TEST_ASSERT( 0 == mbedtls_cipher_update_ad( &ctx, NULL, 0 ) );
+    
+    TEST_ASSERT( 0 == mbedtls_cipher_update( &ctx, buffer0, 1024, buffer0, &buffer0_len ) );
+    
+
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.function]: missing call to `mbedtls_cipher_finish()` and `mbedtls_cipher_reset()`

File: tests/suites/test_suite_cipher.function [hunk 9]
```diff
@@ -712,3 +712,75 @@ void check_padding( int pad_mode, char *input_str, int ret, int dlen_check )
...
+        /* advance pointer into input and output2 to reflect encryption */
+        current += len;
+        buffer1_len += len;
+      
+        /* if buffer1 is all encrypted, break out of loop */
+        if ( current == end )
+	        break;
+    }
+
```
Comment (RonEld) [inline tests/suites/test_suite_cipher.function]: missing call to `mbedtls_cipher_finish()`

File: library/gcm.c (hunk 6 above)
Comment (hanno-arm) [reply library/gcm.c]: Please replace all uses of tabs by whitespaces.

**Comment by Bizyroth** [comment]
@hanno-arm @RonEld 
Hello
I had to replace the crypto part from the [libsrtp](https://github.com/cisco/libsrtp). The cisco libsrtp use OpenSSL as a crypto library and i wanted to use mbedtls to be coherent in my project.

I achieve to replace the crypto but when i run unit test from libsrtp, it failed. After debugging, i noticed that the behavior of mbedtls was different from OpenSSL.

Large buffer was cutted into many smaller buffer (to allow transfer to be efficient over the network) before encryption. The smaller buffer was not cutted into buffer with a size of a multiple of 16 bytes. 

When the receiver receives this large buffer, he cuts it into small buffer too but not at the same location that the sender did. Here is a quick example:
I want to encrypt a buffer of 16 bytes:

1. With OpenSSL:
- I have a context ctx and i encrypt the first 8 bytes from my buffer with the function encrypt(ctx, buffer[0:8])
- I encrypt the last 8 bytes from my buffer with encrypt(ctx, buffer[8:16]). OpenSSL remember that at the last call of encrypt, he didnt see 16 bytes and dont have update the context.
- So if i decrypt then the first 3 bytes with decrypt(ctx, buffer[0;3]) and the rest with decrypt(ctx, buffer[,3,16]) decryption will worked

2. With mbedtls:
- I encrypt the first 8 bytes with encrypt(ctx, buffer[0:8])
- Then when i encrypt the rest of the buffer, the context was update even if i didnt see 16 bytes
- Afterthat when i do decrypt(ctx, buffer[0:3]) and decrypt(ctx, buffer[3:16]) decryption will failed because context used to encrypt buffer[3:8] was different of the context used to decrypt buffer[3:8].

To resume, context in mbedtls are automatically update at the end of a mbedtls_gcm_update (and its written in the documentation, so this behavior is expected) but i think we can add some information to the context to remember if the last buffer we update was not a multiple of 16 bytes to keep the context until we saw 16 bytes.

I hope my explanation are clear, if you want precision i will be able to answer you. I make the change requested soon as i can.

**Comment by Bizyroth** [comment]
@RonEld 
I just saw your comment about unprocessed_data in cipher_ctx. I didnt see this parameter and i think it will be a better approach to solve my problem. It is use with aes_CBC and aes_CTR but not with aes_GCM. I have to check that before i can have an opinion.

**Comment by RonEld** [comment]
Superseded by #1731 so closing this PR.
If you believe this PR was mistakenly closed, please reopen.

**Pull request closed without merging**